contracts/solana/programs/intent/src/instructions/recv_message.rs

```diff
@@ -0,0 +1,146 @@
+use anchor_lang::{
+    prelude::*,
+    solana_program::{instruction::Instruction, program::invoke_signed},
+};
+
+use crate::{
+    connection,
+    error::IntentError,
+    helpers,
+    misc::*,
+    state::*,
+    types::{
+        order_cancel::Cancel,
+        order_fill::OrderFill,
+        order_message::{MessageType, OrderMessage},
+        swap_order::SwapOrder,
+    },
+};
+
+pub fn recv_message<'info>(
+    ctx: Context<'_, '_, '_, 'info, RecvMessageCtx<'info>>,
+    src_network: String,
+    msg: Vec<u8>,
+) -> Result<()> {
+    connection::recv_message(&mut ctx.accounts.receipt, ctx.bumps.receipt)?;
+
+    let order_msg = OrderMessage::try_from(&msg)?;
+    match order_msg.message_type() {
+        MessageType::FILL => {
+            let fill = OrderFill::try_from(&order_msg.message())?;
+            let order = SwapOrder::try_from(&fill.order_bytes())?;
+            invoke_resolve(
+                src_network.clone(),
+                Some(fill),
+                None,
+                order,
+                &ctx.accounts.signer,
+                &ctx.accounts.system_program,
+                ctx.remaining_accounts,
+                &ctx.accounts.config.to_account_info(),
+                &[&[Config::SEED_PREFIX.as_bytes(), &[ctx.accounts.config.bump]]],
+            )
+        }
+        MessageType::CANCEL => {
+            let cancel = Cancel::try_from(&order_msg.message())?;
+            let order = SwapOrder::try_from(&cancel.order_bytes())?;
+            invoke_resolve(
+                src_network.clone(),
+                None,
+                Some(cancel),
+                order,
+                &ctx.accounts.signer,
+                &ctx.accounts.system_program,
+                ctx.remaining_accounts,
+                &ctx.accounts.config.to_account_info(),
+                &[&[Config::SEED_PREFIX.as_bytes(), &[ctx.accounts.config.bump]]],
+            )
+        }
+    }
+}
+
+pub fn invoke_resolve<'info>(
+    src_network: String,
+    fill: Option<OrderFill>,
+    cancel: Option<Cancel>,
+    order: SwapOrder,
```

Comment:
instead of doing optional Cancel and Fill , better way would be to declare enum that has either fill or cancel data because they are mutually exclusive options. That way you can just do match on invoke resolve without any worry.